### PR TITLE
Bump jCommander to 1.82

### PR DESCRIPTION
**Pull request by ukengineer925**
Relates to BG#1468.

### CODE REVIEW

**Review by cfkoehler — APPROVED** [review]

File: src/main/java/emissary/command/HelpCommand.java [hunk 1]
```diff
@@ -36,20 +36,24 @@ public String getCommandName() {
...
         } else {
-            String subcommand = getSubcommand();
-            LOG.info("Detailed help for: " + subcommand);
+            final String subcommand = getSubcommand();
+            LOG.info("Detailed help for: {}", subcommand);
             try {
-                jc.usage(subcommand);
-            } catch (ParameterException e) {
-                LOG.error("ERROR: invalid command name: " + subcommand);
+                // TODO: Once jCommander 1.83 or higher is released, swap back to jc#usage(subcommand)...
```
Comment (jpdahlke) [inline src/main/java/emissary/command/HelpCommand.java:50]: @dev-mlb mentioned that this flip-flop was a thing but he wasn't sure when 1.83 would actually get released. Is it worth waiting? The other updates in the file seem fine.

Comment (ukengineer925) [reply src/main/java/emissary/command/HelpCommand.java:50]: Just circling back on this, 1.83 still hasn't been released yet and there aren't any signs of it coming any time soon.

**Comment by ukengineer925** [comment]
@jpdahlke could we get this lined up for a future release? There's still been no movement on a new version of jCommander, would be nice to get this boxed off soon.

**Comment by cfkoehler** [comment]
@ukengineer925 I did a rebase and pushed up the merges of the import section. Hope you don't mind. Figured I left you waiting long enough I can help with the PR maintenance.

**Review by cfkoehler — DISMISSED** [review]

**Comment by cfkoehler** [comment]
The help command now gives an exception if a subcommand is missing:
```
sh emissary help
 ________  ________ _____ _____  ___  ________   __
|  ___|  \/  |_   _/  ___/  ___|/ _ \ | ___ \ \ / /
| |__ | .  . | | | \ `--.\ `--./ /_\ \| |_/ /\ V /
|  __|| |\/| | | |  `--. \`--. \  _  ||    /  \ /
| |___| |  | |_| |_/\__/ /\__/ / | | || |\ \  | |
\____/\_|  |_/\___/\____/\____/\_| |_/\_| \_| \_/

Command threw an exception: [help]
com.beust.jcommander.ParameterException: There should be exactly 1 main parameters but 0 were found
	at com.beust.jcommander.JCommander.validateOptions(JCommander.java:415)
	at com.beust.jcommander.JCommander.parse(JCommander.java:364)
	at com.beust.jcommander.JCommander.parseValues(JCommander.java:803)
	at com.beust.jcommander.JCommander.parse(JCommander.java:363)
	at com.beust.jcommander.JCommander.parse(JCommander.java:342)
	at emissary.Emissary.execute(Emissary.java:103)
	at emissary.Emissary.main(Emissary.java:162)
```
I will work on a fix

File: src/main/java/emissary/Emissary.java [hunk 2]
```diff
@@ -101,6 +102,15 @@ protected void execute(String[] args) {
         }
         try {
             jc.parse(args);
+        } catch (MissingCommandException e) {
```
Comment (cfkoehler) [inline src/main/java/emissary/Emissary.java:105]: @ukengineer925 
This was an idea on how to handle the additional verification that Jcommander now does.
Sorry to step on your work. If you have a better idea on how to handle this I am open to it.

**Comment by burritoqueen** [comment]
@ukengineer925 / @cfkoehler do you know what the newest version that would not have the breaking change is? For instance, if we can get to 1.79 or 1.81 that would be a big improvement over what we have now, but not require the flip-flopping.  

Also, there are other changes, looks like sonar related, that are included in this PR that are being blocked on the JCommander update.  So I'd like to either find a version we can live with, or pull those out separately so we can get them through.

We're also discussing the possibility of moving to a different command line parser that is more actively maintained/updated.

**Comment by cfkoehler** [comment]
@burritoqueen I don't know the full history here but is there a reason we are even attempting to upgrade in the first place. I am not seeing anything that we would take advantage of today in the release notes. 
But it is definitely worth checking to see if any version greater then 1.58 and does not break anything.

I have been using (picoli)[https://picocli.info/] for command line parsing but that might be a big change to implement here

**Comment by cfkoehler** [comment]
We can upgrade to 1.64. After that the API changes and modifications to the HelpCommand.java are necessary

**Comment by burritoqueen** [comment]
Fun story, picocli is exactly the one we were discussing.  

The upgrade is just because the version we are using is quite old, a code health thing.  As far as I know, there isn't some new functionality we wanted to take advantage of.  But it would be good to get the other changes that are being blocked in this PR.  I believe we can take the remaining changes without updating jcommander (or just moving to 1.64).  Let's do that for now while the other folks are moving to see if picocli can meet our needs (and if you have info there from your own investigations, feel free to share with them)

**Comment by cfkoehler** [comment]
Closing this PR in favor of a smaller version bump with no API changes:
https://github.com/NationalSecurityAgency/emissary/pull/501

**Pull request closed without merging**